### PR TITLE
treedb: certify typed-column layout contracts

### LINKED ISSUE #1850: Add writer-certified typed-column layout contracts
Parent tracker: #1836
Category: write-time certification / read-hot validation elimination

## Goal

Add **writer-certified typed-column layout contracts** so hot reads can avoid repeated parsing and validation after a prepared session has verified the immutable asset once.

Done means the typed-column write path emits canonical, self-describing, validated layout metadata for direct-view and streaming-fast paths; the read path can validate a small descriptor/manifest contract at prepare time, bind it to an immutable asset ref/checksum, and then run hot scans without rechecking byte alignment, length, endian, compression, row-count, dictionary identity, or section shape on every operation.

## Why This Exists

The fastest disk-bytes-to-values path is not to parse defensively on every hot scan. The writer controls layout and should guarantee that sections are written in a form the reader can certify once:

- fixed-width direct-view sections are aligned, little-endian, contiguous, and length-multiple checked;
- varint/bitpack/dictionary/nullable sections are canonical and have bounded descriptor metadata;
- row counts, block offsets, stored bytes, raw bytes, null/default masks, dictionary identities, and stats sections agree;
- immutable asset checksums/ref identity bind the contract to the bytes.

This allows #1849 to use direct views or streaming reducers without paying full parse/validation costs per hot run.

## Safety Rule

Write-time certification reduces hot-read overhead; it does **not** mean blindly trusting bytes forever.

The required trust boundary is:

1. Writer builds canonical bytes and validates invariants before commit/checkpoint.
2. Asset metadata records a versioned layout contract and checksums/lengths.
3. Reader prepare validates contract version, immutable asset ref, descriptor/manifest, checksum policy, and any required section identities once per prepared session/ref.
4. Hot scans reuse the certified plan/views while the session pins the resource lifetime.
5. Corrupt, stale, mismatched, unsupported, or old-format assets fail closed or fall back as explicitly designed.

## Desired State

Writer emits enough information for readers to certify fast paths cheaply:

- logical value type and physical typedcolumn type;
- encoding and compression;
- section category and operation role;
- element size and required alignment for direct-view-capable sections;
- endian contract;
- row count, block row spans, and payload byte lengths;
- fixed-width dimension/degree for vectors/adjaency;
- null/default mask presence and row counts;
- dictionary identity/order/collation contract where relevant;
- stats/pruning metadata applicability;
- checksums/section lengths/descriptor stored bytes;
- flags/capabilities declaring whether direct-view, streaming reducer, stats shortcut, or pruning shortcut may be used after prepare-time certification.

## Scope

- Define writer-certified typed-column layout contract fields, reusing #1843 semantics and #1838 layout capabilities.
- Update typed-column builders/image writers to align/pad direct-view-capable sections and validate invariants before writing.
- Add prepare-time certification that produces a session-local token/plan consumed by #1849 direct-view/streaming decode and #1839 kernels.
- Add diagnostics for certification success/failure and fallback reasons.
- Add tests that corrupt offsets, lengths, row counts, encodings, compression flags, dictionary identities, null/default masks, vector dims, and stats bindings.
- Document pre-alpha format expectations; no complex migration scaffolding is required, but old/unsupported assets must fail clearly or fall back safely.

## Non-Goals And Boundary

- Do not skip checksum/integrity policy; move validation to prepare time where allowed, not remove it.
- Do not add hidden global caches of certified layouts.
- Do not direct-view compressed, variable-width, unaligned, wrong-endian, or unsupported sections.
- Do not trust dictionary code ordering as lexical ordering unless the writer records and validates that contract.
- Do not add new public value types here.
- Do not require every current type to support direct view; some codecs should certify streaming reducer paths instead.

## Required Tests

- [ ] Writer validates and records direct-view contracts for aligned fixed-width sections.
- [ ] Writer rejects or canonicalizes unaligned/truncated/wrong-length fixed-width sections.
- [ ] Reopen validates certified contracts and hot scans use the certified plan.
- [ ] Corrupt descriptor stored bytes, section length, offset, row count, encoding, compression, endian/alignment flags fail closed.
- [ ] Dictionary identity/order mismatch prevents string lexical/dictionary fast paths.
- [ ] Nullable/default mask row-count mismatch prevents value fast paths.
- [ ] Vector dimension and adjacency degree mismatches fail closed.
- [ ] Old/no-contract assets fail with clear pre-alpha error or fall back safely as designed.
- [ ] Prepared session close releases any certified direct-view handles.

## Required Benchmarks And Profiles

- [ ] Measure prepare-time certification overhead separately from hot scan time.
- [ ] Before/after int64 prepared-hot raw/fixed-width direct-view scan showing no per-run validation overhead.
- [ ] Delta-varint streaming reducer path showing descriptor certification avoids repeated section parsing.
- [ ] Dense vector/adjaency direct-view access benchmark if those sections are touched.
- [ ] Report direct-view-certified count, streaming-certified count, certification failures/fallback reasons, `ns/op`, `ops/sec`, `B/op`, `allocs/op`, mapped/decoded/materialized bytes.

## Dependencies And Ordering

- Depends on #1843 semantic capability matrix.
- Depends on #1838 layout/codec capability contracts.
- Feeds #1849 fast decode/direct-view layer.
- Feeds #1839 kernel dispatch and #1840 stats shortcuts by giving them certified section contracts.

## Completion Criteria

- [ ] The writer guarantees canonical layout invariants for certified typed-column sections.
- [ ] The reader validates certification once per immutable asset ref/prepared session and avoids repeated hot-run parsing/validation.
- [ ] Direct-view and streaming-fast paths consume certification rather than ad-hoc parsing.
- [ ] Corruption/staleness/old-format behavior is fail-closed or explicitly safe fallback.
- [ ] CI green and AI review findings resolved or documented.

## Optimized CRC32 Verification Substrate Addition

#1851 adds an early checksum substrate ticket: standardize TreeDB CRC32 verification on `github.com/snissn/go-crc32-asm` through audited wrappers, prove IEEE CRC32 compatibility, and benchmark page/node, commit/WAL, value-log, column asset, typed-column, manifest, vector graph, and template checksum paths.

This must preserve the existing on-disk CRC32 format and read-integrity semantics. It accelerates verification; it does not skip checksums or weaken fail-closed corruption behavior.

### PR DESCRIPTION
**Pull request by snissn**
## Summary

Closes #1850. Part of tracker #1836.

This PR adds writer-certified typed-column layout contracts for newly written typed-column part images and wires prepared int64 aggregate sessions to validate those contracts once per immutable asset/session before hot scans.

## Scope / layer

Layer: #1850 writer-certified layout contracts.

Changed:
- `typedcolumn` images now emit a versioned `layout_contract` section.
- The contract records descriptor checksum, section offsets/lengths/checksums, physical/logical layout metadata, fixed-width shape/alignment/endian, null/default and dictionary identity metadata, and direct/streaming/stats/pruning flags.
- Writers validate the generated contract before returning image bytes.
- Prepared int64 aggregate setup reads and certifies the contract, binds it to existing asset checksum/read-integrity behavior, stores the certified plan in caller-owned session state, and raw fixed-width hot scans reuse that plan instead of re-running generic payload layout validation.
- Diagnostics/benchmarks report direct/streaming certification counts and failures.
- Docs updated for pre-alpha format behavior.

Not changed:
- No unsafe zero-copy typed slices (#1849).
- No new public `ColumnStoreValueType`s.
- No hidden global/process-wide certification cache.

## Supported / fallback

- Non-null `int64` raw fixed-width (`fixed_width_encoding: little_endian`) certifies direct-view eligibility metadata and uses the existing safe byte-loop reducer.
- Non-null `int64` delta-varint certifies streaming reducer metadata.
- Dense vector/adjacency sections are represented in the writer contract and validated by typedcolumn tests, but no new vector/adjacency hot path is introduced here.
- Nullable/default wrappers record mask/count metadata but do not become value-fast-path eligible.
- String dictionary contracts record dictionary identity; lexical order remains unsupported without order+collation proof.
- Missing/old/no-contract assets fail closed during prepared certification with a pre-alpha rebuild error.

## Integrity / lifetime boundary

- Existing asset checksum and read-integrity modes are preserved.
- Cached-verify prepared sessions still validate immutable full assets once before targeted range reads.
- Certification is session-owned (`typedColumnPreparedScanState`) and is cleared on `Close`; mappedresource handles remain scoped to the existing prepared session read cache.

## Tests

- `go test ./TreeDB/internal/typedcolumn ./TreeDB/internal/columnlayout ./TreeDB/collections`
- `go test ./TreeDB/...`

Coverage highlights:
- Writer records aligned fixed-width direct-view contracts.
- Contract/descriptor corruption fails closed for row count, descriptor stored bytes/checksum, section length, payload offset, encoding, compression, endian, alignment, vector dimensions, dictionary identity/order, and nullable mask counts.
- Prepared raw int64 reopen/hot scan validates certification and avoids decoded heap copies.
- Prepared session close releases active mapped/heap handles.
- Corrupt stored contract bytes fail closed.

## Benchmarks

Command:

```sh
TREEDB_TYPED_COLUMN_BENCH_ROWS=65536 \
TREEDB_TYPED_COLUMN_BENCH_LAYOUTS=all \
TREEDB_TYPED_COLUMN_BENCH_READ_INTEGRITY=cached_verify \
TREEDB_TYPED_COLUMN_BENCH_SHAPES=no_filter \
TREEDB_TYPED_COLUMN_BENCH_DISTS=clustered_monotonic \
TREEDB_TYPED_COLUMN_BENCH_INCLUDE_FALLBACK=false \
go test ./TreeDB/collections -run '^$' \
  -bench 'BenchmarkTypedColumnInt64(PrepareCertification|PredicateAggregate)' \
  -benchtime=3x -count=1
```

| Benchmark | ns/op | ops/sec | B/op | allocs/op | direct certified | streaming certified | decoded bytes/op | mapped bytes/op | materialized bytes/op | materialized rows/docs |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| prepare delta-varint certification | 524,611 | 4,239 | 48,176 | 172 | 2 | 2 | 0 | 5,576 | 0 | 0 |
| prepare raw-int64 certification | 457,556 | 6,533 | 48,154 | 172 | 4 | 0 | 0 | 5,576 | 0 | 0 |
| one-shot delta no-filter aggregate | 376,097 | 2,659 | 126,709 | 212 | 2 | 2 | 65,550 | 71,126 | 0 | 0 |
| prepared-hot delta no-filter aggregate | 211,570 | 4,727 | 512 | 4 | prepare: 2 | prepare: 2 | 65,550 | 65,550 | 0 | 0 |
| one-shot raw no-filter aggregate | 317,042 | 3,154 | 58,997 | 211 | 4 | 0 | 0 | 529,864 | 0 | 0 |
| prepared-hot raw no-filter aggregate | 159,180 | 6,282 | 512 | 4 | prepare: 4 | prepare: 0 | 0 | 524,288 | 0 | 0 |

Certification failures were `0/op` in all benchmark rows above.

## Risks / deferrals

- The contract section is a pre-alpha on-disk format addition; old assets must be rebuilt for prepared certified fast paths.
- Direct-view metadata is certified, but #1849 still owns any unsafe/zero-copy typed slice exposure.
- Stats/pruning flags are metadata-only here; #1840/#1841 own stats/pruning frameworks.

## AI review / CI

- Internal subagent implementation/review loop: implementation subagent completed substrate patch; xhigh reviewer PASS with no blocking findings.
- Codex/Copilot/CodeRabbit review loop complete: prior feedback addressed on `44b6182c3`; latest Copilot/Codex found no additional changes, and CodeRabbit status is green/skipped after re-review trigger.
- Latest local validation on head `44b6182c3`: `go test ./TreeDB/...` and the benchmark command above passed. Latest CI/checks are green on head `44b6182c3`.

Not merged; coordinator owns merging.

### CODE REVIEW

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review latest head 7cffd28c9

**Comment by snissn** [comment]
@copilot review latest head 7cffd28c9

**Comment by snissn** [comment]
@coderabbitai review latest head 7cffd28c9

**Comment by snissn** [comment]
Addressed latest review feedback in 44b6182c3:\n\n- require descriptor bytes/length in explicit layout-contract certification;\n- require exact logical type match before direct/streaming certification (missing metadata now fails closed to no fast-path certification);\n- propagate unsupported column type errors from layout-contract encoding;\n- make stored_bytes corruption test target a structured encoded-field offset instead of byte-pattern search;\n- populate materialized_bytes for document fallback materializations.\n\nValidation run on latest head:\n- go test ./TreeDB/...\n- benchmark command in PR body (passed, updated metrics).

**Comment by snissn** [comment]
@copilot review latest head 44b6182c3 after addressing prior feedback

**Comment by snissn** [comment]
@coderabbitai review latest head 44b6182c3 after addressing prior feedback

**Comment by snissn** [comment]
@codex review latest head 44b6182c3 after addressing prior feedback

File: TreeDB/internal/typedcolumn/layout_contract.go [hunk 1]
```diff
@@ -0,0 +1,919 @@
...
+	if contract.ManifestBytes != image.ManifestBytes {
+		return cert, fmt.Errorf("typedcolumn: layout contract manifest bytes=%d want %d", contract.ManifestBytes, image.ManifestBytes)
+	}
+	descriptorSection, err := image.singleSection(ColumnPartImageSectionDescriptor)
+	if err != nil {
+		return cert, err
+	}
+	if err := certifyLayoutContractSection("descriptor", contract.Descriptor, descriptorSection, descriptorRaw); err != nil {
+		return cert, err
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/typedcolumn/layout_contract.go:269]: Fixed in 44b6182c3: explicit certification now requires non-nil descriptor bytes and checks descriptor byte length against the descriptor section before checksum validation.

File: TreeDB/internal/typedcolumn/layout_contract.go [hunk 2]
```diff
@@ -0,0 +1,919 @@
...
+	layoutContractFlagDictionary
+	layoutContractFlagDictionaryOrdered
+	layoutContractFlagNullMask
+	layoutContractFlagDefaultMask
+)
+
+func encodeColumnPartLayoutContractColumn(enc *columnPartImageEncoder, column ColumnPartLayoutContractColumn) {
+	enc.str(column.Name)
+	enc.str(column.LogicalType)
+	columnTypeCode, _ := columnTypeCode(column.Type)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/typedcolumn/layout_contract.go]: Fixed in 44b6182c3: layout-contract column encoding now propagates columnTypeCode errors.

File: TreeDB/collections/typed_column_int64_scan.go [hunk 3]
```diff
@@ -83,6 +83,13 @@ type TypedColumnInt64PredicateScanDiagnostics struct {
 	HeapCopyBytes               uint64
 	DecodedMetadataBytes        uint64
 	DecodedHeapCopyBytes        uint64
+	MaterializedBytes           uint64
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/typed_column_int64_scan.go:86]: Fixed in 44b6182c3: document fallback paths now add ID+document bytes to MaterializedBytes before benchmark/diagnostic reporting.

File: TreeDB/internal/typedcolumn/layout_contract_test.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+	// Corrupt the serialized stored_bytes for the first value block in-place to
+	// prove the stored contract bytes, not only decoded helper structs, fail.
+	encoded := encodeLayoutContractForTest(contract)
+	needle := make([]byte, 8)
+	binary.LittleEndian.PutUint64(needle, uint64(value.Blocks[0].StoredBytes))
+	idx := strings.Index(string(encoded), string(needle))
+	if idx < 0 {
+		t.Fatalf("stored bytes needle not found")
+	}
+	binary.LittleEndian.PutUint64(corrupt[section.Offset+idx:section.Offset+idx+8], uint64(value.Blocks[0].StoredBytes-1))
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/typedcolumn/layout_contract_test.go:482]: Fixed in 44b6182c3: the test now walks the contract encoding to the structured block stored_bytes field offset instead of searching for an 8-byte pattern.

File: TreeDB/internal/typedcolumn/layout_contract.go [hunk 5]
```diff
@@ -0,0 +1,919 @@
...
+		return columnPartContractLayout{elementSize: 4, alignment: 4, endian: ColumnPartLayoutEndianLittle, lengthMultiple: 4, direct: direct, pruning: direct}
+	case EncodingDeltaVarint, EncodingDoubleDeltaVarint:
+		streaming := def.Compression == CompressionNone && (logicalType == "" || logicalType == "int64")
+		return columnPartContractLayout{endian: ColumnPartLayoutEndianCodecDefined, streaming: streaming, stats: streaming, pruning: streaming}
+	case EncodingBoolBitpackRLE:
+		streaming := def.Compression == CompressionNone && (logicalType == "" || logicalType == "bool")
+		return columnPartContractLayout{endian: ColumnPartLayoutEndianCodecDefined, streaming: streaming, stats: streaming}
+	case EncodingLowCardinalityUint32:
+		streaming := def.Compression == CompressionNone && (logicalType == "" || logicalType == "string")
+		return columnPartContractLayout{endian: ColumnPartLayoutEndianCodecDefined, streaming: streaming, stats: streaming, pruning: streaming}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/typedcolumn/layout_contract.go:466]: Fixed in 44b6182c3: direct/streaming certification now requires exact logical type matches; missing logical type metadata no longer certifies fast paths, with test coverage added.

**Comment by snissn** [comment]
Coordinator verified latest head `44b6182c3ef856d81ed0763c4fc060bd028134d9` and the previously unresolved review findings:

- `encodeColumnPartLayoutContractColumn` now propagates `columnTypeCode` errors instead of encoding a zero type code.
- The stored-bytes corruption test now computes the structured field offset by walking the contract encoding rather than searching for an 8-byte pattern.
- Document fallback paths now increment `MaterializedBytes`, so benchmark/diagnostic reporting is not permanently zero for materializing fallback work.

Focused coordinator validation also passed:
`go test -count=1 ./TreeDB/internal/typedcolumn ./TreeDB/internal/columnlayout ./TreeDB/collections -run 'TestColumnPartLayoutContract|TestTypedColumnInt64RawLayoutRoundTripReopenQuery|TestTypedColumnInt64PreparedCertificationDiagnosticsAndClose|TestTypedColumnInt64PreparedLayoutContractCorruptionFailsClosed|TestTypedColumnPreparedState|TestTypedColumnPreparedColumnStateAllowsZeroLengthEmptyBlock'`.

Resolving the addressed review threads.
